### DISCUSSION TITLE
Advice around pre-seeding plugin cache

### DISCUSSION THREAD

**Discussion by jmreicha** [gruntwork-io/knowledge-base#442] [Help] [answered]
Attempting to leverage Terraform plugin caching in our Terragrunt environment and running into race conditions with concurrency.

I saw a few workarounds in https://github.com/gruntwork-io/terragrunt/issues/1212 but was wondering if there are other ways around the problem as that issue is nearly a year old and I’m not sure if there have been any new features that could help.

I’m thinking maybe an error_hook and/or some kind of retry but could use some other ideas.

---

<ins datetime="2022-06-01T02:34:02Z">
  <p><a href="https://support.gruntwork.io/hc/requests/108687">Tracked in ticket #108687</a></p>
</ins>

**Comment by yorinasub17** [comment]
Unfortunately, we or the community hasn't really made progress on this so what's stated in the issue is still relevant at this point in time.

**Reply by jmreicha** [reply]
> Continuously cycle between deleting the terragrunt cache (find . -name ".terragrunt-cache" | xargs rm -r) and running terragrunt validate-all until the plugin cache is seeded.

Thanks for the quick response. Is there an example config/resource you could point me towards that implements this?

**Reply by yorinasub17** [reply]
Nope unfortunately I don't have an example script that automates this. This was more meant to be a manual approach.

**Reply by jmreicha** [reply]
Oh I see, ok thanks. Does an error_hook approach sound plausible as a work around for this?

**Reply by yorinasub17** [reply]
`error_hook` doesn't work here because it's not a retry mechanism.

I think https://terragrunt.gruntwork.io/docs/features/auto-retry/ is what you are looking for, which might work but I haven't tried it so I am not sure.

**Reply by jmreicha** [reply]
That’s good info, I will see if I can twist this into working. Appreciate the guidance.

**Reply by jmreicha** [reply]
So it looks like the retry mechanism is working, as I see the failures now multiple times in plan outputs, but it appears I need a way to clear the `.terraform.hcl.lock` and `.terragrunt-cache` directories that get created the first time https://github.com/gruntwork-io/terragrunt/issues/1960.

**Comment by jmreicha** [comment]
Think I'm getting close to a solution with the following, the problem with this is that when you get into a weird state with providers and  run a `terragrunt plan` it never recognizes that it needs to run an `init` again first to grab the providers it needs. Any ideas?

```
terraform {
  extra_arguments "defaults" {
    commands  = ["init"]
    arguments = ["-no-color"]
  }

  extra_arguments "cache" {
    commands = [
      "init",
      "apply",
      "refresh",
      "import",
      "plan",
      "taint",
      "untaint",
      "show"
    ]
    env_vars = {
      TF_PLUGIN_CACHE_DIR = "/tmp/plugins"
    }
  }

  error_hook "clear_cache" {
    commands = ["init", "plan", "apply"]
    execute  = ["rm", "-rf", "/tmp/plugins/registry.terraform.io"]
    on_errors = [
      "Inconsistent dependency lock file",
      "Required plugins are not installed",
      "Failed to install provider from shared cache",
    ]
  }
}

retryable_errors = [
  "Inconsistent dependency lock file",
  "Required plugins are not installed",
  "Failed to install provider from shared cache",
]
```

**Reply by jmreicha** [reply]
Hrm, it looks like this `error_hook` is maybe never being called 🤔

**Reply by jmreicha** [reply]
@yorinasub17 The `on_errors` block doesn't appear to work with normal regex syntax (as with retryable_errors), is this expected? The only syntax I can get working is `".*"` but I can't figure out how to filter for specific errors. Is that how this is supposed to work?

**Reply by jmreicha** [reply]
Just saw this https://github.com/gruntwork-io/terragrunt/pull/2101.

**Comment by jmreicha** [comment, marked as answer]
If anyone else runs across this we were able to get the configuration working in newer Terragrunt versions (v0.36.12+).

```
terraform {
    extra_arguments "cache" {
      commands = [
        "init",
        "apply",
        "refresh",
        "import",
        "plan",
        "taint",
        "untaint",
        "show"
      ]
      env_vars = {
        TF_PLUGIN_CACHE_DIR = "/tmp/plugins"
      }
    }

    error_hook "mirror_providers" {
      commands = ["init", "plan"]
      execute  = ["terragrunt", "providers", "mirror", "/tmp/plugins"]
      on_errors = [
        "Required plugins are not installed",
      ]
    }
  }

  retryable_errors = [
    "Required plugins are not installed",
  ]

  retry_max_attempts = 3
  retry_sleep_interval_sec = 10
  ...
}
```
